### ISSUE TITLE
Unexpected optic generation where ambiguity is expected

### ISSUE THREAD

**Issue by jppellet** [milessabin/shapeless#665]
Given this context:

```scala
sealed trait T
case class A(s: String) extends T
case class B(s: String) extends T
```

this expression `lens[T].s` yields a `Prism[T, String]` that handles `A` values and ignores `B`. A GitHub discussion suggested the generation should have failed due to ambiguity instead.

Other idea: that this expression yields a `Lens[T, String]` that can handle both `A` and `B` values :-)

**Comment by joroKr21** [comment]
I think instead of ambiguity it would be better if this generates a `Lens[T, String]`

**Comment by milessabin** [comment]
I honestly think it would be better to retire shapeless lenses in favour of Monocle.

**Comment by joroKr21** [comment]
> I honestly think it would be better to retire shapeless lenses in favour of Monocle.

Fair point although I'm not sure if Monocle offers the same convenience (e.g. `Dynamic`)
